py/redrock/rebin.py

```diff
@@ -125,24 +239,150 @@ def trapz_rebin(x, y, xnew=None, edges=None):
+    batch_trapz_rebin_kernel = cp_module.get_function('batch_trapz_rebin')
+
+    nbin = cp.int32(len(edges)-1)
+    nz = cp.int32(len(myz))
+    n = nbin*nbasis*nz
+    nt = cp.int32(len(x))
+
+    blocks = (n+block_size-1)//block_size
+    result = cp.empty((nz, nbin, nbasis), dtype=cp.float64)
+
+    #Launch kernel and syncrhronize
+    batch_trapz_rebin_kernel((blocks,), (block_size,), (x, y, edges, myz, idx, result, nz, nbin, nbasis, nt))
+    #cp.cuda.Stream.null.synchronize()
+
+    #Squeeze array and remove dims of 1
+    if (result.size == 1):
+        #Use flatten because squeeze results in array dim 0
+        result = result.flatten()
+    elif (nz == 1 or nbasis == 1):
+        result = result.squeeze()
+    return result
+
+def rebin_template(template, myz, dwave, use_gpu=False):
     """Rebin a template to a set of wavelengths.
 
-    Given a template and a single redshift, rebin the template to a set of
-    wavelength arrays.
+    Given a template and a single redshift - or an array of redshifts,
+    rebin the template to a set of wavelength arrays.
 
     Args:
         template (Template): the template object
-        z (float): the redshift
+        myz (float or array of float): the redshift(s)
         dwave (dict): the keys are the "wavehash" and the values
             are a 1D array containing the wavelength grid.
 
     Returns:
         dict:  The rebinned template for every basis function and wavelength
-            grid in dwave.
+            grid in dwave.  This is a dict of np or cp 3d arrays
+            (nz x nlambda x nbasis)
 
     """
     nbasis = template.flux.shape[0]  #- number of template basis vectors
-    result = { hs:np.array([ trapz_rebin((1.+z)*template.wave, template.flux[b], wave) for b in range(nbasis) ]).transpose() for hs, wave in dwave.items() }
+    result = dict()
+    if (use_gpu):
+        #In GPU mode, rebin all z and all bases in batch in parallel
+        #and return dict of 3-d CUPY arrays
+        for hs, wave in dwave.items():
+            result[hs] = trapz_rebin(template.wave, template.flux, xnew=wave, myz=myz, use_gpu=True)
+    elif (type(myz) == np.ndarray):
+        #If myz is numpy array, process all redshifts and all bases (one at
+        #a time) and collect results in dict of 3-d numpy arrays
+        nz = len(myz)
+        result = { hs:np.empty((nz, len(wave), nbasis)) for hs, wave in dwave.items() }
+        for i, z in enumerate(myz):
+            x = (1.+z)*template.wave
+            for hs, wave in dwave.items():
+                for b in range(nbasis):
+                    result[hs][i,:,b] = trapz_rebin(x, template.flux[b], wave)
+    else:
+        #Legacy mode - a single redshift passed - return a dict with rebinned
+        #template for every basis for this one redshift.
+        result = { hs:np.array([ trapz_rebin((1.+myz)*template.wave, template.flux[b], wave) for b in range(nbasis) ]).transpose() for hs, wave in dwave.items() }
```

Comment:
Currently `rebin_template` with `use_gpu=False` supports scalar or vector `myz`, but `use_gpu=True` only supports vector `myz` and crashes if `myz` is a scalar.  This is another case where were should make the CPU vs. GPU API/feature identical, perhaps my adding support for scalar `myz` directly to `trapz_rebin` (for both `use_gpu=True/False`).

i.e. even if the rest of the GPU code would never call this with a scalar `myz`, for maintainability and PLS(S), every function that supports `use_gpu=True/False` should be API/feature/result identical unless there is a very specific reason not to (e.g. an option like blocksize that only applies to one but not the other).

Reply:
Done